### ISSUE TITLE
Build directory for editable installations

### ISSUE THREAD

**Issue by dnicolodi** [mesonbuild/meson-python#370]
The build directory for editable build is currently `.mesonpy/editable/build` in the source directory. I don't see a reason why the build directory is nested two levels deep into otherwise empty directories . A flatter directory structure could be used. 

While doing so, attaching a Python API tag to the build directory name would allow to have multiple parallel installations of the same package for different interpreters.

Isn't there prior art for choosing the name of the build directory? I don't like something specific to meson-python much. Python uses `__pycache__` for its cache. Is it allowed to add tool specific stuff in there? distutils uses a `build` folder, IIRC. What about other build backends?

It would be nice to have this decided before we release a version with the editable support.

EDIT: reworded for politeness

**Comment by rgommers** [comment]
> Isn't there prior art for choosing the name of the build directory?

I'd say `build/` in the root dir is the one more or less standard name - not just distutils/setuptools, but also Sphinx (also that also uses `_build` sometimes). Not sure about other backends, never had a need to actually use them.

The downside of `build` is that it may clash with a standard choice for using Meson directly (`meson setup build`, `ninja -C build` is already mostly muscle memory for me).

Reusing `__pycache__` is a no no I'd say.

> Also attaching a Python API tag to the build directory name would allow to have multiple parallel installations of the same package for different interpreters.

Hmm, that is an interesting point.

**Comment by rgommers** [comment]
I'd say that whatever we choose here is a private implementation detail though - the user has no business doing anything with it beyond adding it to `.gitignore`, so we should be fine tweaking it later.

**Comment by rgommers** [comment]
Thinking about this some more:
- a hidden dir isn't nice, it's already tricked me once or twice, and I kinda knew it should be there
- `build` is the standard name
- separate interpreters can already be supported via the `builddir` option (I tested it with `pip install -e . --no-build-isolation --config-settings=builddir=build-py310` and it works)
- `build` clashing with the `distutils` default may actually be a good thing; having in-tree remnants from a distutils build can in some cases cause problems

tl;dr let's use `build`?

**Comment by dnicolodi** [comment]
I like `build` too. I like supporting multiple interpreters out of the box. Having a two level directory structure solves this: `build/cp311` or `build/cpython-3.11/` but I don't know if this is a common enough use case that is worth supporting.

**Comment by rgommers** [comment]
I'd prefer to not special-case interpreters I think. Different builds can also happen with different compilers (e.g., using GCC and Clang in parallel), different sets of build dependencies (on SciPy testing with 2 numpy versions is much more interesting than testing with 2 cpython versions), and so on.

Plus it does bring up questions like "what if we have regular and a debug 3.11 interpreters".

Matrix'ed environment testing is for tools like tox/nox or custom scripts I'd say, and they can easily pass a `builddir` for anything like this.

**Comment by FFY00** [comment]
Other possible option would be putting the build directory on the installation path, instead of the project source. This would allow installations to different interpreters without any conflicts.

**Comment by rgommers** [comment]
I'm sorry to see this interaction, as well as the pattern of lots of back and forth in general. It's safe to say that things could go more smoothly in communication on this project. I hope that we all strive to be open, kind and collaborative. If we'd try to be a little less strongly worded in our opinions (myself included), I think that would improve things. We're all pretty competent developers here, but what often seems to be the problem is we missed some use case or have some silent assumption in our heads that turns out to be incomplete. 

Regarding reviews and decisions, it'd be great to distinguish more between what's important/blocking, and what's really not. This issue is a good example: we can make a small improvement in usability here, but it really doesn't matter all that much. And a similar thing applies to any review comments: style comments,  grammar nits, using a set vs. a tuple or other such things that don't change how code works, etc. - let's just try to _not_ comment on those, and if we do indicate that we understand that it's not critical by adding "minor" or "nit"? 

To bring that back to a more neutral example: my last doc PR (gh-338) attracted 44 comments - nothing particularly bad about any single comment, but the overall effect is that it's energy-draining and the end result of the doc content really didn't change all that much compared to the initial version. If we'd just merged it as is and spent the remaining time on other docs work, we'd have accomplished more with less friction.

To address directly the CoC comments: 
- We don't have a CoC right now, and this is not the right time to discuss introducing one. Once things settle down I'd be happy to take the lead on that though, since it's good practice to have one (ideally we'd have all that is listed under [Community Standards](https://github.com/mesonbuild/meson-python/community)). In the meantime, let's all try to be kind/respectful/professional/`<choose your preferred term here>`. 
- I don't see the quoted sentence as a personal attack.
  - It is strongly worded and, as stated above, it'd be useful to tone that down a little. But that goes for all of us, and for quite a few other review interactions.

**Comment by rgommers** [comment]
> > Also attaching a Python API tag to the build directory name would allow to have multiple parallel installations of the same package for different interpreters.
> 
> +1 on this.

Okay, you both agree on this, so let's go that way.

For the other choices, how about I make an editorial decision here? Seems preferred over lots more discussion.

**Comment by dnicolodi** [comment]
I proposed tagging the build directory with the python interpreter API tag in some form because the editable wheel documentation page prominently warns (or at least I remember it this way) to the fact that being able to have editable wheels for multiple interpreters as a limitation. I thought that, if the limitation was felt so important, adding a tiny amount of complexity to allow it could have been a good compromise.

However, I'm a big fan of keeping things simple unless necessary, thus I'm happy without this feature and a paragraph in the documentation that mentions how to achieve the same using the `builddir` config setting.

I'm fine with @rgommers picking the build directory name that he prefers. I opened this issue only because the nested directories are  not required to exist anymore and I thought we could simplify what we present to the users before releasing editable wheel support in the next release.

**Comment by FFY00** [comment]
@dnicolodi thank you for replying.

> I considered for quite a while whether I should respond in the merit of the accusation of not being considerate in my communication. I really wanted to just ignore it: I'm not contributing to open source for the drama and I really don't have emotive energy to sink into this kind of arguments. This is the reason why I didn't even read your email on the subject: I don't have the emotional resources for it.

Me neither. That is why I tried to reach out to try to sort our issues out, as we are both committed to working on this project, so interactions were not going to stop. Ignoring the issue and letting things snowball further will IMO generally only make things worse, which turned out to be the case here.

I dropped the ball on making you feel welcome on this project, leading to our interaction in #334, and that is something I acknowledge and apologize for.

Since things did not get better since my reply there, two weeks later I sent an email (titled "meson-python collaboration") attempting again to figure out how to sort out our issues, but I am not sure what other steps I could/should have taken.

> However, you really seem convinced that my choice of words here and elsewhere is a personal attack toward you, so much that it constitutes a violation of community code of conduct. I cannot let pass this characterization of what I wrote.

To be clear, I do not consider your choice of words in this issue a personal attack. The issue is the approach (no questions, no reasoning, simply asking for an action) coupled with the language, which I've asked you to be more careful with, but this is a very small issue on its own. In hindsight, this was a bad place to elevate the communication issue further.

Regarding the CoC discussion, this is mostly a non-issue standalone. I brought the CoC up because I believe your other messages, in addition to the fact that you haven't adjusted the language in these situations after I asked you to, can be indeed a PSF CoC violation. While we haven't adopted that CoC, we do work closely with many projects that use it, so I thought it was relevant to mention as a warning.

An example of some of the other messages I mention here is https://github.com/mesonbuild/meson-python/pull/359#discussion_r1145980026, which does seem to me like a personal attack. A more constructive reply could be something like "Aren't we dealing with sysconfig concepts? In this context, wouldn't "scheme" be the collections of concrete directories where the parts of a wheel are installed?". To which I would have explained that I was referring to wheel schemes (see https://github.com/pypa/installer/blob/f89b5d93a643ef5e9858a6e3f450c83a57bbe1f1/src/installer/utils.py#L82 for eg.), given that we are building a wheel, not installing on the file-system or anything that has to interface with `sysconfig`[^1]. Your interpretation is still reasonable – this whole ecosystem is a mess anyway – but the assumption that it was so obvious that it warranted you commenting on my understanding of the concept, and even bringing up my involvement in `sysconfig`, is simply not right.

[^1]: Other than the fact that we used to have to compare the paths Meson wants to install to with the `sysconfig` paths, as part of the heuristics. That is conceptually detached IMO, though.

While my communication isn't the best, I don't think I have ever made such a statement regarding you, or any other kind of attack. Your complaints seem to have been due to me not reviewing your PRs in a timely manner and submitting a PR with a similar purpose (in #334), which I apologized for on the issue and via email, and you feeling that I am purposely disagreeing with you (in #360, #358), which I can understand how you would feel like that, even if I am not doing anything on purpose. I am not sure if there's anything else, but these are the only ones I remember you vocalized.

Maybe I am being too naive, or misguided, but I think these are things that shouldn't be very hard to squash if you were open to, and that would significantly improve our further interactions, hence my attempts at doing so. Text-based communication is very hard afterall.

> Can you please tell me what you find offensive in me pointing out that a technical minutia does not serve anymore any purpose? Would the use of the words "do not serve anymore any purpose" instead of "are kind of useless" have made a difference in your interpretation of my statement? I'm clearly not a native English speaker, but AFAIK, "useless" just means "not fulfilling the intended purpose" and does not carry any judgement on who put the object in place. What if the word "wrong" would have been used instead? Would you consider stating that something is wrong or broken or that does not function as intended a stab at who implemented it?
> 
> It did not even cross my mind to give any judgment on who coded the solution in this way. We continuously make decision that will be revealed to be wrong down the road. If this would not be the case, software would be without bugs. I don't think that pointing out that (with more information and in mutated circumstanced) some of these decisions appear wrong should be seen as a judgment of the person that made these decisions.
> 
> In this specific instance, I know why the directory structure was chosen to be what it is. It was a decision that made sense when it was made. However, with mutated circumstances (the new implementation of the editable wheels) this specific directory structure does not serve anymore the intended purpose. I think it is not controversial to state that, because there is no more installation directory for editable wheels, having the `build` component in the `.mesonpy/editable/build/` path does not have any use: : it is kind of useless.

I wasn't clear in my previous reply, so I apologize for that. I'll try to be as clear as possible here: I don't think those directories are useless, I don't think I ever documented the reasoning behind using them, so I don't think you know all the motivations behind the decision, especially since you state that you know they used to make sense at the time, but don't anymore, as I don't think anything that major changed. The problem is not that you think the directories are useless, it's that you simply state that, while not providing any reasoning or what changed for that to be the case. I believe this is kind of a nuanced thing, because we are always compromising between too much and too little verbosity, but unless you are absolutely certain about something or are providing reasoning, I'd avoid using terms like "useless". This is another thing that could simply be rewritten as a question, making it much better, but for bigger decisions, we should also consider stating the previous status quo, what changed, what that means, and the proposed action.

> This kind of statements bring a technical discussion to the personal level and are not useful to the consensus building process. In particular, the "you don't understand" argument not followed by an explanation is so bad that I feel it is almost stereotypical of bad arguments. It inherently describes one side of the discussion as being right and the other wrong: it reads pretty much as "I have the knowledge and you don't. I'm going to dismiss everything else you have to say because the product of ignorance. Once you will inform yourself, you will inevitably agree with me". Please note how this logically reduces to "You don't understand till you don't agree with me", which is dangerously close to "I am right, and you are wrong". Consider this a warning for your future interactions with fellow human beings: this kind of arguments tend to do not be received wheel.

That is fair. I wasn't trying to get into the technical details of the issue, but given that I made that statement, I should have elaborated. I hope my reasoning is clearer now, as I explain above that I don't understand how you can be so sure about the considerations behind the decision.

---

Anyway, the way we communicate/interact is not healthy, and I am not sure how that can actually be improved at this point, which is fine, it is what it is. Not everyone works well together, so let's not keep deteriorating both our mental health. I resign as a meson-python maintainer, which I think should make things much easier for both of us. Due to my job, I'll likely still be contributing to the project, but I won't be making the final calls or be responsible for the decisions, so I should be able to just state my recommendations and let y'all make the calls. I am gonna to remove myself from the PyPI and readthedocs pages, and the `meson-python-admins` and `meson-python-devs` Github teams, staying only on the `meson-python-triage` one, but, naturally, feel free to change that if you want. Hopefully, things will quiet down now.

@dnicolodi I thank you again for your contributions to this project so far, and wish you all the best going forward. I am sorry for how things turned out between us, and hope that with time tensions can die down.

**Comment by FFY00** [comment]
I must still be leftover with commit permissions in the repo settings directly. @rgommers if you could remove that it'd be great, thanks.

**Comment by dnicolodi** [comment]
This is exactly the drama I'm not here for. You take everything I write as personally directed to you. I guarantee you that this is not the case: I have no issue in working with you. There have been a few occasions in which what you wrote, and how wrote it, made me angry. Debate has inevitably moments like this, even when all sides are the most considerate. When that happens I go away, possibly take a walk, get calm, and try my best to  respond only on the technical merits. If when I come back I still see an issue, I've let you know and considered the incident closed. You have excuses for these instances of poor behavior from your side, but you seem to treat all instances of my behavior that made you angry as a mounting attack toward you.

I don't like that your poor communication choices, including instances in which you have been purposely disrespectful and others where you didn't take the time to read and understand what was discussed before proclaiming resolutions, should be excused but you keep escalating any minor issue you find in my communication. I don't think that assuming that everything I write has the goal of attacking you personally and continuously escalating the handling of a situation that I guarantee you does not exist are honest attempts to resolve a conflict.

One example at hand is the thread in this issue. I have been harsh and definitive in describing a technical solution that you adopted. I understand that you may have not liked that. However, instead that pointing out the reasons why my description of your technical solution was incorrect and you decided to neglect the technical side of the issue and focus on the form I used to describe the issue. We are several messages into the discussion and there is still no explanation of what is the purpose of the directory structure as it is. Have you started your reply with that, I would have been prompt to apologize for describing it as useless. However, you decided to escalate the issue you saw with the form of my writing and go on with the very poor "you don't understand" argument. A simple "I don't like the way you decided to word this" would have also been a valid response, that most likely would have resulted in an apology and us moving forward with other things. Assuming intention, connecting it to a pattern of inconsiderate behavior, and treating it as a CoC sanctionable offense, on the other hand, is guaranteed to trigger a defensive response.

I've explained that what I wrote was not intended to be judgemental or disrespectful and I asked for an explanation of what you found offensive in it. As a result you decided to escalate again and step down as a maintainer. I respect your decision, I hope this will help you to see things under a different prospective, and that it will make you happier. I wish there would have been another way for you to deal with the situation. I apologize for anything I wrote that made you feel personally attacked. It was not my intention.

**Comment by dnicolodi** [comment]
> I was referring to wheel schemes (see https://github.com/pypa/installer/blob/f89b5d93a643ef5e9858a6e3f450c83a57bbe1f1/src/installer/utils.py#L82 for eg.), given that we are building a wheel, not installing on the file-system or anything that has to interface with `sysconfig`

I don't think that a poorly named variable in a source file from a third project is a relevant source for a naming convention. If you look at the normative PEP linked in the line above the one you are linking to, or to the PyPA document that replaced it, here https://packaging.python.org/en/latest/specifications/binary-distribution-format/#file-contents you will find that what the folders installed into a wheel archive are referred to as "install scheme keys", which is reasonable to unpack to mean "keys into the scheme that defines the installation paths". Nowhere I see the folders themselves defines as "schemes". As "keys" is pretty generic, but using the word "paths" as used by `sysconfig` is completely unambiguous. On the contrary, using the word "scheme" is confusing, because it has no precedent for being used in this was in the PEPs and PyPA normative documents on which we base out implementation. It is of very little relevance how another project decided decided to name its internal variables.

**Comment by FFY00** [comment]
I always try to be charitable, and I have done that with you. I don't know how it was from your side, but my perception was that everything was okay up to #334, but since then you started being more aggressive in replies (this is not necessarily an issue, everyone has their communication style, but I thought this as a deterioration in our communication), started making what IMO are clearly personal attacks (eg. https://github.com/mesonbuild/meson-python/pull/304#discussion_r1128705367, https://github.com/mesonbuild/meson-python/pull/359#discussion_r1145980026), and started merging PRs with unresolved review comments after arguments (eg. #280, #279).

This issue was the last drop. I had already asked you to be more careful with the way you word things (see https://github.com/mesonbuild/meson-python/pull/289#issuecomment-1468766909), so after reading it and having gone through the last month with the issues I mentioned above, I decided to escalate the situation. It was a poor decision to do it here, and the way I did it was also poor, as I have mentioned in my previous replies.

You say you wish there would have been another way for me to deal with this situation, but I don't know what you expect me to have done, because I did try. After I was seemingly unsuccessful at de-escalating the situation in #334, I sent you an email (titled "meson-python collaboration"), to which I did not get any reply, and you now say you ignored. After that, I mentioned our communication being unhealthy in https://github.com/mesonbuild/meson-python/pull/358#discussion_r1145969680 and asked if you had seen my email, to which you replied that there was no communication issue and that I sometimes simply do not seem to handle disagreement very well. Which, I mean, can be a valid criticism, but the example you provide is me giving up on a technical discussion right after you attack me on another PR (https://github.com/mesonbuild/meson-python/pull/359#discussion_r1145980026). I don't know how I am supposed to keep being charitable after those kinds of attacks, or why I am expected to keep going with technical arguments following such interactions, but most of all, I don't understand how anything warrants that kind of response. If you didn't think there was a communication issue at that time, then do you think the personal attacks, etc., are normal and acceptable communication? After all this, I gave up and escalated the issue.

As I said before, I think a lot of this snowballed from misunderstandings between us and our communication styles. Those kinds of things are not uncommon, as communication, especially text-based communication, is difficult, so it is my opinion this is something you can't really expect always being able to avoid. What I think is super important, especially between co-maintainers in a project, is being able to recover from such situations and resolve any issues as soon as possible, instead of letting them snowball, which I tried to do.

Anyway, I just want to be done with this. I accept your apology and hope we can move past this, but I do not think the issues raised here are fully resolved. I think a good way forward would be for the project maintainers to have the discussion about what conduct is and isn't accepted, setting the same expectations for everybody. Hopefully we both can take this as a lesson, and use it to avoid similar situations in the future.

**Comment by dnicolodi** [comment]
> I'd say that whatever we choose here is a private implementation detail though - the user has no business doing anything with it beyond adding it to `.gitignore`, so we should be fine tweaking it later.

We can use the same trick that Meson uses and place a `.gitignore` containing `*` in the build directory. Meson does not do that when the build directory is not empty, and we add the native file to the build directory before invoking `meson setup`, thus Meson does not do that for us.

**Comment by dnicolodi** [comment]
setuptools uses a `build/__editable__.*` directory for editable installs in strict mode https://setuptools.pypa.io/en/latest/userguide/development_mode.html#strict-editable-installs I haven't checked yet what `.*` should be replaced with
